### PR TITLE
Add recursive create/delete

### PR DESCRIPTION
**Pull request by jiacheliu3**
### What changes are proposed in this pull request?

The delete/create options are currently not passed to the UFS. By passing them to the UFS correctly, we should be able to achieve recursive create/delete.

However, note that under Dora structure, when we delete `/a/`, we can't also remove all `/a/*` cache from all workers immediately. https://github.com/Alluxio/alluxio/issues/17741

### CODE REVIEW

**Comment by jiacheliu3** [comment]
I did try to write a UT but the one below doesn't work. If someone can enlighten me on how a proper UT can be wrote, I figure I can add some proper coverage to what I'm changing here.
```
  @Test
  public void testCreateFile() throws Exception {
    File testDir = mTestFolder.newFolder("testDir");
    testDir.mkdirs();
    File testFile = new File(testDir, "a");

    OpenFileHandle handle = mWorker.createFile(testFile.getPath(),
        CreateFilePOptions.getDefaultInstance());
    // PagedDoraWorker.completeFile complains the file does not exist in UFS
    mWorker.completeFile(testFile.getPath(), CompleteFilePOptions.getDefaultInstance(),
        handle.getUUID().toString());

    // My intentions is the file should have been created by the worker in UFS
    assertTrue(testFile.exists());
    FileInfo fileInfo = mWorker.getFileInfo(testFile.getPath(), null);
    System.out.println(fileInfo);
  }
```

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 1]
```diff
@@ -737,7 +741,9 @@ private void invalidateFileMeta(String path) {
     } else {
       parentDir = fullPathUri.getParent();
     }
-    mListStatusCache.invalidate(parentDir.toString()); // invalidate dir cache
+    if (parentDir != null) {
+      mListStatusCache.invalidate(parentDir.toString()); // invalidate dir cache
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:745]: avoid NPE

**Review by huanghua78 — APPROVED** [review]

**Comment by huanghua78** [comment]
> I did try to write a UT but the one below doesn't work. If someone can enlighten me on how a proper UT can be wrote, I figure I can add some proper coverage to what I'm changing here.
> 
> ```
>   @Test
>   public void testCreateFile() throws Exception {
>     File testDir = mTestFolder.newFolder("testDir");
>     testDir.mkdirs();
>     File testFile = new File(testDir, "a");
> 
>     OpenFileHandle handle = mWorker.createFile(testFile.getPath(),
>         CreateFilePOptions.getDefaultInstance());
>     // PagedDoraWorker.completeFile complains the file does not exist in UFS
>     mWorker.completeFile(testFile.getPath(), CompleteFilePOptions.getDefaultInstance(),
>         handle.getUUID().toString());
> 
>     // My intentions is the file should have been created by the worker in UFS
>     assertTrue(testFile.exists());
>     FileInfo fileInfo = mWorker.getFileInfo(testFile.getPath(), null);
>     System.out.println(fileInfo);
>   }
> ```


The problem in current implementation is:
worker does not create the target file.
The target file is created by client (client writes to UFS directly).

So the UT would not work as expected.

Solution:
```

 @Test
  public void testCreateFile() throws Exception {
    File testDir = mTestFolder.newFolder("testDir");
    testDir.mkdirs();
    File testFile = new File(testDir, "a");

    OpenFileHandle handle = mWorker.createFile(testFile.getPath(),
        CreateFilePOptions.getDefaultInstance());
    // create file and write some data directly to this file in UFS. 
    int length = 1024;
    byte[] buffer = BufferUtils.getIncreasingByteArray(length);
    BufferUtils.writeBufferToFile(testFile.getAbsolutePath(), buffer);

    mWorker.completeFile(testFile.getPath(), CompleteFilePOptions.getDefaultInstance(),
        handle.getUUID().toString());

    assertTrue(testFile.exists());
    FileInfo fileInfo = mWorker.getFileInfo(testFile.getPath(),
        GetStatusPOptions.getDefaultInstance());
    assertEquals(fileInfo.getLength(), length);
    System.out.println(fileInfo);
  }

```

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**